server/src/services/stack.service.ts

```diff
@@ -0,0 +1,77 @@
+import { BadRequestException, Inject, Injectable } from '@nestjs/common';
+import { AccessCore, Permission } from 'src/cores/access.core';
+import { BulkIdsDto } from 'src/dtos/asset-ids.response.dto';
+import { AuthDto } from 'src/dtos/auth.dto';
+import { StackCreateDto, StackResponseDto, StackSearchDto, StackUpdateDto, mapStack } from 'src/dtos/stack.dto';
+import { IAccessRepository } from 'src/interfaces/access.interface';
+import { ClientEvent, IEventRepository } from 'src/interfaces/event.interface';
+import { IStackRepository } from 'src/interfaces/stack.interface';
+
+@Injectable()
+export class StackService {
+  private access: AccessCore;
+
+  constructor(
+    @Inject(IAccessRepository) accessRepository: IAccessRepository,
+    @Inject(IEventRepository) private eventRepository: IEventRepository,
+    @Inject(IStackRepository) private stackRepository: IStackRepository,
+  ) {
+    this.access = AccessCore.create(accessRepository);
+  }
+
+  async search(auth: AuthDto, dto: StackSearchDto): Promise<StackResponseDto[]> {
+    const stacks = await this.stackRepository.search({
+      ownerId: auth.user.id,
+      primaryAssetId: dto.primaryAssetId,
+    });
+
+    return stacks.map((stack) => mapStack(stack, { auth }));
+  }
+
+  async create(auth: AuthDto, dto: StackCreateDto): Promise<StackResponseDto> {
+    await this.access.requirePermission(auth, Permission.ASSET_READ, dto.assetIds);
+
+    const stack = await this.stackRepository.create({ ownerId: auth.user.id, assetIds: dto.assetIds });
+
+    this.eventRepository.clientSend(ClientEvent.ASSET_STACK_UPDATE, auth.user.id, dto.assetIds);
+
+    return mapStack(stack, { auth });
+  }
+
+  async get(auth: AuthDto, id: string): Promise<StackResponseDto> {
+    await this.access.requirePermission(auth, Permission.STACK_READ, id);
+    const stack = await this.findOrFail(id);
+    return mapStack(stack, { auth });
+  }
+
+  async update(auth: AuthDto, id: string, dto: StackUpdateDto): Promise<StackResponseDto> {
+    await this.access.requirePermission(auth, Permission.STACK_WRITE, id);
```

Comment:
Assuming you add new assets to the stack, you'd probably also want to enforce the same permissions for them you'd enforce when creating a stack

Reply:
I believe update only supports changing the primary asset id, which already has to be in the stack, so I don't think any checks are necessary about asset visibility.